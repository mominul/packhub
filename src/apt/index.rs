use std::{collections::BTreeMap, io::Write};

use anyhow::Result;
use askama::Template;
use chrono::{DateTime, Utc};
use libflate::gzip::{EncodeOptions, Encoder, HeaderBuilder};
use md5::Md5;
use sha1::Sha1;
use sha2::{Sha256, Sha512};

use crate::{
    apt::deb::DebianPackage,
    package::Package,
    utils::{Arch, ReleaseChannel, hashsum},
};

#[derive(Debug)]
pub struct AptIndices {
    packages: BTreeMap<Arch, Vec<DebianPackage>>,
    date: DateTime<Utc>,
}

#[derive(Template)]
#[template(path = "Release")]
struct ReleaseIndex<'a> {
    channel: &'a ReleaseChannel,
    arch: Vec<String>,
    date: String,
    files: Vec<Files>,
}

#[derive(Template)]
#[template(path = "Packages")]
struct PackageIndex<'a> {
    packages: &'a [DebianPackage],
}

struct Files {
    md5: String,
    sha1: String,
    sha256: String,
    sha512: String,
    size: usize,
    path: String,
}

impl AptIndices {
    pub fn new(packages: &[Package]) -> Result<AptIndices> {
        let mut debian: BTreeMap<Arch, Vec<DebianPackage>> = BTreeMap::new();
        // Find the latest date from the list of packages
        let mut date = DateTime::UNIX_EPOCH;
        for package in packages {
            if *package.creation_date() > date {
                date = *package.creation_date();
            }

            match DebianPackage::from_package(package) {
                Ok(deb) => {
                    if let Some(arch) = deb.get_arch() {
                        debian.entry(arch).or_default().push(deb);
                    } else {
                        tracing::error!(
                            "Debian package architecture not found for package: {:?}",
                            package
                        );
                        continue;
                    }
                }
                Err(e) => {
                    tracing::error!("Error occurred when extracting debian control data: {e}");
                    continue;
                }
            }
        }
        Ok(AptIndices {
            packages: debian,
            date,
        })
    }

    pub fn get_package_index(&self, arch: &Arch) -> String {
        let index = PackageIndex {
            packages: self.packages.get(arch).unwrap(),
        };
        index.render().unwrap().trim().to_owned()
    }

    pub fn get_release_index(&self, channel: &ReleaseChannel) -> String {
        let date = self.date.to_rfc2822();

        let mut files = vec![];

        for arch in self.packages.keys() {
            let packages = self.get_package_index(arch);
            let packages = packages.as_bytes();
            let packages_gz = gzip_compression(packages);

            files.extend([
                Files {
                    sha256: hashsum::<Sha256>(packages),
                    size: packages.len(),
                    path: format!("main/binary-{}/Packages", arch),
                    md5: hashsum::<Md5>(packages),
                    sha1: hashsum::<Sha1>(packages),
                    sha512: hashsum::<Sha512>(packages),
                },
                Files {
                    sha256: hashsum::<Sha256>(&packages_gz),
                    size: packages_gz.len(),
                    path: format!("main/binary-{}/Packages.gz", arch),
                    md5: hashsum::<Md5>(&packages_gz),
                    sha1: hashsum::<Sha1>(&packages_gz),
                    sha512: hashsum::<Sha512>(&packages_gz),
                },
            ]);
        }

        // Sort the files
        files.sort_by(|a, b| a.path.cmp(&b.path));

        let index = ReleaseIndex {
            date,
            files,
            channel,
            arch: self.packages.keys().map(|s| s.to_string()).collect(),
        };

        index.render().unwrap()
    }
}

pub fn gzip_compression(data: &[u8]) -> Vec<u8> {
    let header = HeaderBuilder::new().modification_time(0).finish();
    let options = EncodeOptions::new().header(header);
    let mut encoder = Encoder::with_options(Vec::new(), options).unwrap();
    encoder.write_all(data).unwrap();

    let gzip = encoder.finish();

    gzip.into_result().unwrap()
}

#[cfg(test)]
mod tests {
    use std::fs::{self, read};

    use chrono::DateTime;
    use insta::assert_snapshot;

    use super::*;
    use crate::package::tests::package_with_ver;

    #[test]
    fn test_apt_indices() {
        let package = Package::detect_package("OpenBangla-Keyboard_2.0.0-ubuntu20.04.deb", "2.0.0".to_owned(), "https://github.com/OpenBangla/OpenBangla-Keyboard/releases/download/2.0.0/OpenBangla-Keyboard_2.0.0-ubuntu20.04.deb".to_owned(), DateTime::parse_from_rfc2822("Wed, 8 Nov 2023 16:40:12 +0000").unwrap().into()).unwrap();
        let data = read("data/OpenBangla-Keyboard_2.0.0-ubuntu20.04.deb").unwrap();
        package.set_package_data(data);

        let packages = vec![package];

        let indices = AptIndices::new(&packages).unwrap();

        // Packages
        let packages = indices.get_package_index(&Arch::Amd64);
        assert_snapshot!(packages);

        // Release
        let release = indices.get_release_index(&ReleaseChannel::Stable);
        assert_snapshot!(release);
    }

    #[test]
    fn test_multiple_packages() {
        let package1 = package_with_ver("fcitx-openbangla_3.0.0.deb", "3.0.0");
        let data = fs::read("data/fcitx-openbangla_3.0.0.deb").unwrap();
        package1.set_package_data(data);

        let package2 = package_with_ver("ibus-openbangla_3.0.0.deb", "3.0.0");
        let data = fs::read("data/ibus-openbangla_3.0.0.deb").unwrap();
        package2.set_package_data(data);

        let packages = vec![package1, package2];

        let indices = AptIndices::new(&packages).unwrap();

        // Packages
        let packages = indices.get_package_index(&Arch::Amd64);
        assert_snapshot!(packages);
        assert_eq!(packages.as_bytes().len(), 2729);
        let packages_gz = gzip_compression(packages.as_bytes());
        assert_eq!(packages_gz.len(), 1105);

        // Release
        let release = indices.get_release_index(&ReleaseChannel::Stable);
        assert_snapshot!(release);
    }

    #[test]
    fn test_multiple_architectures() {
        let package1 = package_with_ver("fastfetch-linux-aarch64.deb", "2.40.3");
        let data = fs::read("data/fastfetch-linux-aarch64.deb").unwrap();
        package1.set_package_data(data);

        let package2 = package_with_ver("fastfetch-linux-amd64.deb", "2.40.3");
        let data = fs::read("data/fastfetch-linux-amd64.deb").unwrap();
        package2.set_package_data(data);

        let package3 = package_with_ver("fastfetch-linux-armv6l.deb", "2.40.3");
        let data = fs::read("data/fastfetch-linux-armv6l.deb").unwrap();
        package3.set_package_data(data);

        let package4 = package_with_ver("fastfetch-linux-armv7l.deb", "2.40.3");
        let data = fs::read("data/fastfetch-linux-armv7l.deb").unwrap();
        package4.set_package_data(data);

        let package5 = package_with_ver("fastfetch-linux-ppc64le.deb", "2.40.3");
        let data = fs::read("data/fastfetch-linux-ppc64le.deb").unwrap();
        package5.set_package_data(data);

        let package6 = package_with_ver("fastfetch-linux-riscv64.deb", "2.40.3");
        let data = fs::read("data/fastfetch-linux-riscv64.deb").unwrap();
        package6.set_package_data(data);

        let package7 = package_with_ver("fastfetch-linux-s390x.deb", "2.40.3");
        let data = fs::read("data/fastfetch-linux-s390x.deb").unwrap();
        package7.set_package_data(data);

        let packages = [
            package1, package2, package3, package4, package5, package6, package7,
        ];

        let indices = AptIndices::new(&packages).unwrap();

        // Release
        let release = indices.get_release_index(&ReleaseChannel::Stable);
        assert_snapshot!(release);
    }
}
