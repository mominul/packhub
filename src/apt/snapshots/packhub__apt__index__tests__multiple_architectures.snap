---
source: src/apt/index.rs
expression: release
---
Origin: . stable
Label: . stable
Suite: stable
Codename: stable
Date: Thu, 1 Jan 1970 00:00:00 +0000
Architectures: amd64 arm64 armhf riscv64 s390x
Components: main
Description: Generated by packhub
MD5Sum:
 dc34cc78e7ea211c16648cf4d5058b7f     825      main/binary-amd64/Packages
 1a9cbea2e2af0769beb96c646d6800b9     550      main/binary-amd64/Packages.gz
 00cc0450632472e7d052e1adb546c1bc     827      main/binary-arm64/Packages
 11c7ee672e6e5b44a4cf020806404915     553      main/binary-arm64/Packages.gz
 75de33a4ab436ced20b0d1e5549469f1     1654      main/binary-armhf/Packages
 42c55998d67940fc56a5ee3a98af0ba2     757      main/binary-armhf/Packages.gz
 9b9b742e803f55894c2a6620d1186a07     829      main/binary-riscv64/Packages
 31a3d0b27d0423acf874bf54d7b3d0bd     553      main/binary-riscv64/Packages.gz
 2d2d98bc14b6861107bb75aa9951d534     772      main/binary-s390x/Packages
 6c48949d483864dd964dd9638d5e22f4     525      main/binary-s390x/Packages.gz
SHA1:
 89dfdd9fdaa1bfd8f751cae2fbe173ddeb3bc155     825      main/binary-amd64/Packages
 9a5a5187251f3e13875856c26c94f1caf0dc9938     550      main/binary-amd64/Packages.gz
 391b818f94fb1b814df168b1e1de7616f9e931bd     827      main/binary-arm64/Packages
 b1af467abebba695db6b38a548edd6f9ee21cdbb     553      main/binary-arm64/Packages.gz
 d0ebaf02685f8c37c27ec5e3500a7ab27e334e04     1654      main/binary-armhf/Packages
 7ad164d41bd44745322a577b487785227aef2316     757      main/binary-armhf/Packages.gz
 daf93234659323df286b6782b78260cae5080573     829      main/binary-riscv64/Packages
 61f5ba94a9ea9a5e906e967bb312e05c2eb787dd     553      main/binary-riscv64/Packages.gz
 1ff724ce2104db2f44b28c4b4f418310b9679b0f     772      main/binary-s390x/Packages
 deeec27a71e6de3df724b40120d47f6b8c560e5f     525      main/binary-s390x/Packages.gz
SHA256:
 a77f55baac76d7c608c148cb44d3ebd2729e91b1bc6ce00760e3ab147b4c15e3     825      main/binary-amd64/Packages
 931540243041e730a57498d6fd6e75bf86a3bc1e24feb0950a6373915a298d8a     550      main/binary-amd64/Packages.gz
 20e70c475e8f9fbf9a95209186d2260467b7eec6ffe10ef769b97cb351d6c8d3     827      main/binary-arm64/Packages
 b7dc4f77f370332e3e02004b6b52d2b46f176ce9883114323ffdb2e4fd8ed48b     553      main/binary-arm64/Packages.gz
 ecec52b2cb2072a8f9c56ad57b59f226a8b3889f82824cd52e491d508c842c4b     1654      main/binary-armhf/Packages
 511aeacdeb5a80ef4d3c0d8704c71c71b8c1da0cd6ff8e44397ffee3d9417bea     757      main/binary-armhf/Packages.gz
 5d92bad48122c49c2360030f8752bd94d5ed17b7eb5b8eea17804b08bce9fa7c     829      main/binary-riscv64/Packages
 02e0553328327a7d4a2a76fb8aa3281365780358db95670988b46358b05d7605     553      main/binary-riscv64/Packages.gz
 0163c0b615421b497de5741f2fb2edaa6c5e2d208439a9e3fba077da183df6df     772      main/binary-s390x/Packages
 813299364721e0b57bb7550564ac10bdf457f10a98d79c3c85f3aad8c94009a4     525      main/binary-s390x/Packages.gz
SHA512:
 df7dbdf6b502cb4c5d6b0d25456ea82538c1c0f9d5efb5c618e3341da3ca30e3d18a017f9f73bc7bb6efcb22d5fd0f88d710d14420a77d50d66085fc662c3fef     825      main/binary-amd64/Packages
 1381840a14d3a35fa648e3be35a50f03d07d61e462579fc0f3418e48ccfb71fdc97b7497c7e4d4bcaff2b5d946603a191dbb0362f67fe25e1cbcf38654a41a7f     550      main/binary-amd64/Packages.gz
 ae2422d17499a57d8d37bc4320f8c6b0862ec26b27c5d87f98fba88d899f6c3a6e2e78c95cf7417641c48410560560740a115882d8b6bf68d7413bc0b33dfc0a     827      main/binary-arm64/Packages
 b969bbc6b1d5fcd7e380f402653c0e7ace7b2209d4a89d55dfc2e1ef3f5dad109e61a236eadbd518422986baf0a8a8cc06e9ad1de01bb352344ce9fdd714dadc     553      main/binary-arm64/Packages.gz
 33242dc0c71ecca24ffab2e05d09f5aca172088e7d33ef4005a08334933d7405dbd7ed32d143b4fbe9e8764fb73d64f17a4e890c391963c2043a30e729cf991e     1654      main/binary-armhf/Packages
 b43d1ddb37548cc65f090b72b05163c1203da327717bab930cac3f9fa20ad4bf65113cf4839cabbcfad001a74efec67ac41653bd65caa0024405150d5248b22a     757      main/binary-armhf/Packages.gz
 41c97a30246a2eba40a00db51e4f3cd02dff6b4ea43dc90209170d42a382b9db02fbaf25598e3ff665120e1fc8f62f8692fd1701e0542c6cfe918c4ef5c180d8     829      main/binary-riscv64/Packages
 62e5275713626233f45323ad094a1dadc539ddea3b383ee988ce70691d87249373235162613f6634c3dc20bb4d3fb71facdbaa0b194895a89f2a717a166031f3     553      main/binary-riscv64/Packages.gz
 b8e34ef8efea71ae3a7e3861491c11833d7de76eb41e9f1a3b81d2bb884c914ce6ded05909935892845a9378fa08d0262fbe6cf946832371a5be540f7f38ce13     772      main/binary-s390x/Packages
 c94de85906e90d95af7f5d2cf593721bf79e12ef84dd2cc3ae8250992b65693680be95d989ff659d6a25c360f5a37b7ef5f984f069947e9ee6f8d373a58a1999     525      main/binary-s390x/Packages.gz
